lib/src/config_fetcher.dart

```diff
@@ -1,82 +1,201 @@
+
+  final ConfigCatLogger logger;
+  final ConfigJsonCache jsonCache;
   final String mode;
   final String sdkKey;
-  late final bool urlIsCustom;
-  late final String url;
-
-  ConfigFetcher({
-    required this.log,
-    required this.sdkKey,
-    required this.mode,
-    required DataGovernance dataGovernance,
-    baseUrl = "",
-  }) {
-    this.urlIsCustom = !baseUrl.isEmpty;
-    this.url = baseUrl.isEmpty ? dataGovernance.url : baseUrl;
+
+  late final bool _urlIsCustom;
+  late final Dio _client;
+  late String _etag = '';
+  late String _url;
+
+  ConfigFetcher(this.logger, this.sdkKey, this.mode, this.jsonCache,
+      ConfigCatOptions options) {
+    _urlIsCustom = options.baseUrl.isNotEmpty;
+    _url = _urlIsCustom
+        ? options.baseUrl
+        : options.dataGovernance == DataGovernance.global
+            ? globalBaseUrl
+            : euOnlyBaseUrl;
+
+    _client = Dio(BaseOptions(
+        connectTimeout: options.connectTimeout.inMilliseconds,
+        receiveTimeout: options.receiveTimeout.inMilliseconds,
+        sendTimeout: options.sendTimeout.inMilliseconds,
+        responseType: ResponseType.plain,
+        validateStatus: (status) =>
+            status != null && (status >= 200 && status < 600)));
+
+    if (options.httpClientAdapter != null) {
+      _client.httpClientAdapter = options.httpClientAdapter!;
+    }
+
+    if (options.proxyUrl.isNotEmpty &&
+        _client.httpClientAdapter is DefaultHttpClientAdapter) {
+      (_client.httpClientAdapter as DefaultHttpClientAdapter)
+          .onHttpClientCreate = (client) {
+        client.findProxy = (uri) {
+          return 'PROXY ${options.proxyUrl}';
+        };
+      };
+    }
   }
 
-  Future<FetchResponse> fetchConfigurationJson(http.Client client) {
-    return _executeFetch(client, executionCount: 2);
+  /// Gets the underlying [Dio] client.
+  @override
+  Dio get client {
+    return _client;
   }
 
-  Future<FetchResponse> _executeFetch(http.Client client, {int executionCount = 1}) async {
-    Map<String, String> headers = {
-      'X-ConfigCat-UserAgent': 'ConfigCat-Dart/$mode-$_version',
-    };
+  /// Fetches the current ConfigCat configuration json.
+  @override
+  Future<FetchResponse> fetchConfiguration() {
+    // ensure fetch requests are not initiated simultaneously
+    return syncFuture(() => _executeFetch(2));
+  }
+
+  /// Closes the underlying http connection.
+  @override
+  void close() {
+    _client.close();
+  }
 
-    if (!etag.isEmpty) {
-      headers['If-None-Match'] = etag;
+  Future<FetchResponse> _executeFetch(int executionCount) async {
+    final response = await _doFetch();
+    if (!response.isFetched || response.config?.preferences == null) {
+      return response;
     }
 
-    final response = await client.get(
-      Uri.parse('$url/configuration-files/$sdkKey/$configJsonName.json'),
-      headers: headers,
-    );
+    final preferences = response.config!.preferences!;
```

Comment:
I'm pretty sure we can get rid of those force-unwraps (`!`) by doing just this small tweak:
```suggestion
    // here preferences is nullable
    final preferences = response.config?.preferences;
    if (!response.isFetched || preferences == null) {
       return response;
     }

    // now it's not nullable anymore =]
```

Reply:
👍